lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -304,126 +305,180 @@ void LocalEnforcer::execute_actions(
+          break;
+        }
+        for (const auto& session : it->second) {
+          if (session->get_session_id() == session_id) {
+            auto& uc = session_update[imsi][session_id];
+            start_session_termination(imsi, session, true, uc);
+            break;
+          }
+        }
+        // Session not found
+        MLOG(MERROR) << "Cannot act on TERMINATE action since session "
+                     << session_id << " does not exist";
+        break;
+      }
+      case CONTINUE_SERVICE:
+        break;
     }
   }
 }
 
-// Terminates sessions that correspond to the given IMSI.
-// (For session termination triggered by sessiond)
-void LocalEnforcer::terminate_service(
-    SessionMap& session_map, const std::string& imsi,
-    const std::vector<std::string>& rule_ids,
-    const std::vector<PolicyRule>& dynamic_rules,
-    SessionUpdate& session_update) {
-  pipelined_client_->deactivate_flows_for_rules(
-      imsi, rule_ids, dynamic_rules, RequestOriginType::GX);
-  MLOG(MINFO) << "Initiating session termination for " << imsi;
-
-  auto it = session_map.find(imsi);
-  if (it == session_map.end()) {
-    MLOG(MWARNING) << "Could not find session with IMSI " << imsi
-                   << " to terminate";
+// Terminates sessions that correspond to the given IMSI and session.
+void LocalEnforcer::start_session_termination(
+    const std::string& imsi, const std::unique_ptr<SessionState>& session,
+    bool notify_access, SessionStateUpdateCriteria& update_criteria) {
+  auto session_id = session->get_session_id();
+  if (session->is_terminating()) {
+    // If the session is terminating already, do nothing.
+    MLOG(MDEBUG) << "Session " << session_id << " is already terminating, "
+                 << "ignoring termination request";
     return;
   }
+  MLOG(MINFO) << "Initiating session termination for " << session_id;
 
-  for (const auto& session : it->second) {
-    if (session->is_terminating()) {
-      // If the session is terminating already, do nothing.
-      continue;
+  remove_all_rules_for_termination(imsi, session, update_criteria);
+
+  session->set_fsm_state(SESSION_TERMINATING_FLOW_ACTIVE, update_criteria);
+  auto config = session->get_config();
+  if (notify_access) {
+    notify_termination_to_access(imsi, session_id, config);
+  }
+  if (config.common_context.rat_type() == TGPP_WLAN) {
+    MLOG(MDEBUG) << "Deleting UE MAC flow for subscriber " << imsi;
+    auto mac_addr = config.mac_addr;
+    auto sid      = config.common_context.sid();
+    pipelined_client_->delete_ue_mac_flow(sid, mac_addr);
+  }
+  if (terminate_on_wallet_exhaust()) {
+    handle_subscriber_quota_state_change(
+        imsi, *session, SubscriberQuotaUpdate_Type_TERMINATE, update_criteria);
+  }
+  // The termination should be completed when aggregated usage record no
+  // longer
+  // includes the imsi. If this has not occurred after the timeout, force
+  // terminate the session.
+  evb_->runAfterDelay(
```

Comment:
Do you know if this delay really necessary? Terminating as soon as we receive the termination request we may lose few seconds of tracking but it makes the system really unpredictable. I thin amount of data that is going to be lost is negligible (unless we set pipelined reports to minutes).

Reply:
Yeah I think this waiting logic needs some love, but I think I will handle that in another diff. Eventually, I will have to simplify the terminating_* states for the upcoming FSM state changes, so I'll give it some more thought there.